### DISCUSSION TITLE
Return queried items when using aggregation query (graphql)

### DISCUSSION THREAD

**Discussion by 8byr0** [directus/directus#9424] [Feature Requests]
It's currently not possible to get more than computed values when using `<collection>_aggregated` which is a bit annoying since you need to query another time to retrieve items.

Here would be an ideal query
```graphql
 items_aggregated(groupBy:"project"){
    group
    sum {
         duration
    }
    items {
      id
      duration
      name
    }
  }
```

And its result
```json
{
  "data": {
    "items_aggregated": [
      {
        "group": {
          "project": 1
        },
        "sum": {
          "duration": 50
        },
        "items": [
          {
            "id": 1,
            "duration": 10,
            "name": "random name"
          },
          {
            "id": 2,
            "duration": 40,
            "name": "another name"
          }
        ]
      },
      {
        "group": {
          "project": 2
        },
        "sum": {
          "duration": 72
        },
        "items": [
          {
            "id": 3,
            "duration": 32,
            "name": "abcde"
          },
          {
            "id": 4,
            "duration": 28,
            "name": "fghij"
          },
          {
            "id": 5,
            "duration": 12,
            "name": "klmno"
          }
        ]
      }
    ]
  }
}
```

From what I've seen and tested so far, updating the schema is quite easy to implement (the first snippet is actually a one generated by directus, only the result is fake since it returned null for items in my test) but before going further and starting working on a PR, I need to know what do you think of such feature? Does it make sense? 

(hello @rijkvanzanten, yet another gql discussion 😄 )

**Comment by rijkvanzanten** [comment]
> but before going further and starting working on a PR, I need to know what do you think of such feature? Does it make sense?

I'd love to have it _however_ it won't be easy to implement, as that's not a thing you can easily do in SQL. Currently, the GQL statement is converted like (dummy code):

```graphql
query {
  articles_aggregated(groupBy: "category") {
    avg {
      rating
    }
  }
}
```

```sql
SELECT AVG(rating) FROM articles GROUP BY category;
```

In a query like that there's no way to retrieve the actual rows that were grouped on. In order to implement something like this, we'd have to extract what all the groups are, and then do another query to fetch all the items in those groups which is non-trivial. (cc SQL wizard @Oreilles)

**Reply by 8byr0** [reply]
> In a query like that there's no way to retrieve the actual rows that were grouped on. In order to implement something like this, we'd have to extract what all the groups are, and then do another query to fetch all the items in those groups which is non-trivial. (cc SQL wizard @Oreilles)

Ah damn, totally missed this point (too many mongo developments recently^^).
Is it actually a problem to make another SQL query? Does not sound like a technical pain to implement IMHO but I trust your experience with dynamic queries building.

Another option could be to have a new gql query explicitly for grouping, something like `<item>_item_grouped` which could look like this:
```graphql
query {
<item>_item_grouped(groupBy: "field", filter: ...) {
    group
    items {
        key
	anotherkey
    }
}
```

It may not mean a different backend handling but it may avoid messing up the Aggregation part and keep concerns separated

**Reply by rijkvanzanten** [reply]
Yeah the problem becomes that you have to select those items with a WHERE clause to extract them per group, and then apply limits and sub filters per group, which isn't that easy to do in a single query :) that being said, it's a very similar situation as the solution to fetching nested relational o2m items, so we've done the trick before. I think we can revisit this once we start looking at cleaning up the ast-runner and query-applier

**Reply by 8byr0** [reply]
> I think we can revisit this once we start looking at cleaning up the ast-runner and query-applier

I'd be glad to help then. Do you have an ETA for this cleanup?

**Reply by rijkvanzanten** [reply]
No ETA. Can't promise any timelines on this, as it's a totally new feature request, that hasn't come up before. We're currently a little swamped in improving the general stability of the platform so will circle back to new features like this at a later point 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
